gobblin-modules/gobblin-http/src/main/java/gobblin/http/HttpRequestBuilder.java

```
@@ -0,0 +1,106 @@
+import org.apache.http.client.methods.HttpUriRequest;
+import org.apache.http.client.methods.RequestBuilder;
+import org.apache.http.entity.ContentType;
+import org.apache.http.entity.StringEntity;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import gobblin.utils.HttpUtils;
+import gobblin.writer.http.AsyncWriteRequest;
+import gobblin.writer.http.AsyncWriteRequestBuilder;
+import gobblin.writer.http.BufferedRecord;
+
+
+/**
+ * Build {@link HttpUriRequest} that can talk to http services
+ *
+ * <p>
+ *   This basic implementation builds a write request from a single record. However, it has the extensibility to build
+ *   a write request from batched records, depending on specific implementation of {@link #buildWriteRequest(Queue)}
+ * </p>
+ */
+public class HttpRequestBuilder implements AsyncWriteRequestBuilder<GenericRecord, HttpUriRequest> {
+  private static final Logger LOG = LoggerFactory.getLogger(HttpRequestBuilder.class);
+  private final String urlTemplate;
+  private final String verb;
+
+  public HttpRequestBuilder(String urlTemplate, String verb) {
+    this.urlTemplate = urlTemplate;
+    this.verb = verb;
+  }
+
+  @Override
+  public AsyncWriteRequest<GenericRecord, HttpUriRequest> buildWriteRequest(Queue<BufferedRecord<GenericRecord>> buffer) {
+    return buildWriteRequest(buffer.poll());
+  }
+
+  /**
+   * Build a write request from a single record
+   */
+  private AsyncWriteRequest<GenericRecord, HttpUriRequest> buildWriteRequest(BufferedRecord<GenericRecord> record) {
+    if (record == null) {
+      return null;
+    }
+
+    AsyncWriteRequest<GenericRecord, HttpUriRequest> request = new AsyncWriteRequest<>();
+    HttpOperation httpOperation = HttpUtils.toHttpOperation(record.getRecord());
+
+    // Set uri
+    URI uri = HttpUtils.buildURI(urlTemplate, httpOperation.getKeys(), httpOperation.getQueryParams());
+    if (uri == null) {
+      return null;
```

When the url syntax is invalid. Looks like we should throw an error.